docs/proposals/20210716-async-azure-resource-creation-deletion.md

```diff
@@ -0,0 +1,320 @@
+    LoadBalancersReadyCondition clusterv1.ConditionType = "LoadBalancersReady"
+    // PrivateDNSReadyCondition means the private DNS exists and is ready to be used.
+    PrivateDNSReadyCondition clusterv1.ConditionType = "PrivateDNSReady"
+    // BastionHostReadyCondition means the bastion host exists and is ready to be used.
+    BastionHostReadyCondition clusterv1.ConditionType = "BastionHostReady"
+    // InboundNATRulesReadyCondition means the inbound NAT rules exist and are ready to be used.
+    InboundNATRulesReadyCondition clusterv1.ConditionType = "InboundNATRulesReady"
+    // AvailabilitySetReadyCondition means the availability set exists and is ready to be used.
+    AvailabilitySetReadyCondition clusterv1.ConditionType = "AvailabilitySetReady"
+    // RoleAssignmentReadyCondition means the role assignment exists and is ready to be used.
+    RoleAssignmentReadyCondition clusterv1.ConditionType = "RoleAssignmentReady"
+
+    // CreatingReason means the resource is being created.
+    CreatingReason = "Creating"
+    // FailedReason means the resource failed to be created.
+    FailedReason = "Failed"
+    // DeletingReason means the resource is being deleted.
+    DeletingReason = "Deleting"
+    // DeletedReason means the resource was deleted.
+    DeletedReason = "Deleted"
+    // DeletionFailedReason means the resource failed to be deleted.
+    DeletionFailedReason = "DeletionFailed"
+)
+```
+
+### <a name='OpenQuestions'></a>Open Questions
+
+#### <a name='Whatshouldthetimeoutdurationsbe'></a>1. What should the timeout durations be?
+
+The specific numbers are not set in stone, and should be revised after doing some performance testing with different values and calculating the P99 expected completion time of operations that are not long-running.
+
+The other question is whether we should have the same timeout value for all operations (the 5s) or curate per operation. For simplicity, the proposal is to start with a single value. Later on, we might want to optimize by calculating a dynamic timeout value for each operation based on heuristics. That would be better than statically defining artificial timeout durations for each operation which might vary over time and might not be the same depending on region, subscription, etc.
+
+#### 2. How to handle transient errors in logs?
+
+The idea of short-circuiting the Reconcile loop when a long-running operation is in progress involves returning an error when an operation is not done. This means that the Reconcile loop will end in an error every time an operation is in progress. This is necessary because we need to requeue so that the reconcile loop can run again to check on the progress of the operation, but it also means that the user might will see the error message in the logs. How can we handle transient errors in logs without spamming the logs and therefore causing noise that reduces the user's ability to see actual errors in reconcile?
```

Comment:
Why are we returning an error? Is this because it's a well-established implementation pattern that we trust? It's not clear why we would *have* to represent a non-terminal response state as an error.

Reply:
it's just the way we handle transient errors right now https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/controllers/azuremachine_controller.go#L299 it gets logged as an "error" which could be misleading in this case since a resource not being done provisioning is not an error. I think I have a good idea about how to avoid that though and make sure "still ongoing" requeues don't get logged as errors.